app/src/main/java/com/firebase/uidemo/database/realtime/RealtimeDbChatActivity.java

```diff
@@ -0,0 +1,177 @@
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.Query;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+
+/**
+ * Class demonstrating how to setup a {@link RecyclerView} with an adapter while taking sign-in
+ * states into consideration. Also demonstrates adding data to a ref and then reading it back using
+ * the {@link FirebaseRecyclerAdapter} to build a simple chat app.
+ * <p>
+ * For a general intro to the RecyclerView, see <a href="https://developer.android.com/training/material/lists-cards.html">Creating
+ * Lists</a>.
+ */
+public abstract class RealtimeDbChatActivity extends LifecycleActivity
+        implements FirebaseAuth.AuthStateListener {
+    private static final String TAG = "RealtimeDatabaseDemo";
+
+    /**
+     * Get the last 50 chat messages.
+     */
+    protected static final Query sChatQuery =
+            FirebaseDatabase.getInstance().getReference().child("chats").limitToLast(50);
+
+    @BindView(R.id.messagesList)
+    RecyclerView mRecyclerView;
+
+    @BindView(R.id.sendButton)
+    Button mSendButton;
+
+    @BindView(R.id.messageEdit)
+    EditText mMessageEdit;
+
+    @BindView(R.id.emptyTextView)
+    TextView mEmptyListMessage;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_chat);
+        ButterKnife.bind(this);
+
+        mRecyclerView.setHasFixedSize(true);
+        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+
+        ImeHelper.setImeOnDoneListener(mMessageEdit, new ImeHelper.DonePressedListener() {
+            @Override
+            public void onDonePressed() {
+                onSendClick();
+            }
+        });
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        if (isSignedIn()) { attachRecyclerViewAdapter(); }
+        FirebaseAuth.getInstance().addAuthStateListener(this);
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        FirebaseAuth.getInstance().removeAuthStateListener(this);
+    }
+
+    @Override
+    public void onAuthStateChanged(@NonNull FirebaseAuth auth) {
+        mSendButton.setEnabled(isSignedIn());
+        mMessageEdit.setEnabled(isSignedIn());
+
+        if (isSignedIn()) {
+            attachRecyclerViewAdapter();
+        } else {
+            Toast.makeText(this, R.string.signing_in, Toast.LENGTH_SHORT).show();
+            auth.signInAnonymously()
+                    .addOnSuccessListener(this, new OnSuccessListener<AuthResult>() {
+                        @Override
+                        public void onSuccess(AuthResult result) {
+                            attachRecyclerViewAdapter();
```

Comment:
I think this success handler is not necessary, since when we sign in successfully the listener will trigger and isSignedIn will be true and then we'll attach the adapter.

Reply:
Whoops, refactor mistake. 😄